{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "b7ba7cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import json\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import sys\n",
    "import argparse\n",
    "import re\n",
    "from itertools import permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "911419b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "class EvaluateT1k:\n",
    "    \n",
    "    def __init__(self, inpath = '', gene = 'HLA', \n",
    "                 gtrue = '', outpath = '', method = 'T1K', mode = ''):\n",
    "    \n",
    "        '''\n",
    "        gene = HLA, KIR\n",
    "        method: T1K, arcasHLA, Optitype\n",
    "        mode: ingnore\n",
    "        '''\n",
    "        self.path = inpath\n",
    "        self.gene = gene\n",
    "        self.gtrue = gtrue\n",
    "        self.gtruth = {}\n",
    "        self.samples = []\n",
    "        self.outpath = outpath\n",
    "        self.qualityThreshold = 0\n",
    "        self.method = method\n",
    "        self.mode = mode\n",
    "        self.total_sum = defaultdict(list)\n",
    "        self.samplesOut = defaultdict(list)\n",
    "        \n",
    "        if gene == 'HLA':\n",
    "            self.header = ['sample', 'HLA-A', 'HLA-B','HLA-C', 'HLA-DRB1', 'HLA-DQB1']\n",
    "        \n",
    "        elif gene == 'KIR':\n",
    "            self.header = ['sample', 'KIR3DL2']\n",
    "        \n",
    "    def readDirs(self):\n",
    "        print(\"Initializing current directories...\")\n",
    "        for file in os.listdir(self.path):\n",
    "            fpath = self.path + file\n",
    "            if os.path.isdir(fpath):\n",
    "                self.samples.append(file)\n",
    "        \n",
    "        print(\"Found \", len(self.samples), \" number of sample output!\", )\n",
    "    \n",
    "    def setGtruth(self):\n",
    "        print(\"Reformatting and Storing ground truth...\")\n",
    "        with open(self.gtrue, 'r') as gt:\n",
    "            for line in gt:\n",
    "                cols = line.rstrip().split()\n",
    "                sname = cols[0]\n",
    "                self.gtruth[sname] = {}\n",
    "                for i in range(1, len(self.header)):\n",
    "                    gene = self.header[i]\n",
    "                    \n",
    "                    # missing the ground truth of this gene                    \n",
    "                    if len(cols) <= i:\n",
    "                         self.gtruth[sname][gene] = []\n",
    "                    \n",
    "                    else:\n",
    "                        #complete allele names are constructed\n",
    "                        alleles = cols[i].split('/')\n",
    "                        self.gtruth[sname][gene] = ['.']*len(alleles)\n",
    "                        #Ignore 003 mode\n",
    "                        if self.mode == 'ignore':\n",
    "                            if '003' in alleles:\n",
    "                                #print('003 found in sample')\n",
    "                                self.gtruth[sname][gene] = []\n",
    "                            else:\n",
    "                                nalleles = [gene + \"*\" + a for a in alleles]\n",
    "                                self.gtruth[sname][gene] = nalleles\n",
    "                        elif 'DL1' in alleles:\n",
    "                            #print('DL1 found in sample')\n",
    "                            self.gtruth[sname][gene] = []\n",
    "                        else:\n",
    "                            nalleles = [gene + \"*\" + a for a in alleles]\n",
    "                            self.gtruth[sname][gene] = nalleles\n",
    "    \n",
    "    def ParseAlleleName(self, n):\n",
    "        cols = n.split(\"*\")\n",
    "        if (self.gene == \"HLA\"):\n",
    "            subcols = cols[1].split(\":\")\n",
    "            return cols[0], cols[0] + \"*\" + \":\".join(subcols[0:3]), n\n",
    "        else:\n",
    "            return cols[0], cols[0] + \"*\" + cols[1][0:3], n\n",
    "        \n",
    "    def getMaxAllele(self, genotype, tgenotype):\n",
    "        '''\n",
    "        genotype = ['001', '003']\n",
    "        tgenotype = ['001', '002', '003']\n",
    "        '''\n",
    "\n",
    "        maxScore = 0\n",
    "        maxAllele = ''\n",
    "        for i in permutations(tgenotype,2):\n",
    "            alleleP = [a for a in i]\n",
    "            score = sum([1 if (genotype[r] == alleleP[r]) else 0 for r in range(2)])\n",
    "            if score >= maxScore:\n",
    "                maxScore = score\n",
    "                maxAllele = alleleP\n",
    "        \n",
    "        return maxAllele\n",
    "    \n",
    "    def matchAlleles(self):\n",
    "        \n",
    "        print(\"Start matching alleles...\")\n",
    "        for sname in self.samples:\n",
    "            genotype = {}\n",
    "            allGenes = {}\n",
    "            if self.method == \"T1K\":\n",
    "                fgenotype = os.path.join(self.path, sname,sname) + '_genotype.tsv'\n",
    "            elif self.method == \"arcasHLA\":\n",
    "                fgenotype = os.path.join(self.path, sname,sname) + '.genotype.json'\n",
    "            elif self.method == \"Optitype\":\n",
    "                fgenotype = os.path.join(self.path, sname,sname) + '_result.tsv'\n",
    "            if sname not in self.gtruth:\n",
    "                continue\n",
    "            else:\n",
    "                tgenotype = self.gtruth[sname]\n",
    "            \n",
    "            with open(fgenotype, 'r') as fh:\n",
    "                for line in fh:\n",
    "                    cols = line.rstrip().split()\n",
    "                    if self.method == 'T1K':\n",
    "                        gene = cols[0]\n",
    "                        if (gene not in genotype):\n",
    "                            genotype[gene] = [\".\", \".\"]\n",
    "                        if (gene not in allGenes):\n",
    "                            allGenes[gene] = 1\n",
    "                        \n",
    "                        allele_1 = \".\"\n",
    "                        if (len(cols) >= 5):\n",
    "                            allele_1_s = cols[2].split(\":\")[0:2]\n",
    "                            allele_1_n = \":\".join(allele_1_s)\n",
    "                            allele_1 = re.sub('[A-Za-z]+$', \"\", allele_1_n)\n",
    "                            if (float(cols[4]) > 0):\n",
    "                                genotype[gene][0] = allele_1\n",
    "\n",
    "                        if (len(cols) >= 8):\n",
    "                            allele_2_s = cols[5].split(\":\")[0:2]\n",
    "                            allele_2_n = \":\".join(allele_2_s)\n",
    "                            allele_2 = re.sub('[A-Za-z]+$', \"\", allele_2_n)\n",
    "                            if (float(cols[7]) > 0):\n",
    "                                genotype[gene][1] = allele_2\n",
    "                            if (float(cols[7]) == -1):\n",
    "                                genotype[gene][1] = allele_1\n",
    "                        else:\n",
    "                            genotype[gene][1] = allele_1\n",
    "                    \n",
    "                    elif self.method == 'arcasHLA':\n",
    "                        patterns = re.findall('\\\".*?\\\"', line)\n",
    "                        for p in patterns:\n",
    "                            if \"*\" not in p:\n",
    "                                continue\n",
    "                            idx = 1\n",
    "                            allele = p[1:-1]\n",
    "                            n = str(self.gene) + \"-\" + allele\n",
    "                            gene, majorAllele, allele = self.ParseAlleleName(n)\n",
    "                            if (gene not in genotype):\n",
    "                                genotype[gene] = [\".\", \".\"]\n",
    "                                idx = 0\n",
    "                            if (gene not in allGenes):\n",
    "                                allGenes[gene] = 1\n",
    "                            allele_1_s = majorAllele.split(\":\")[0:2]\n",
    "                            allele_1_n = \":\".join(allele_1_s)\n",
    "                            allele_1 = re.sub('[A-Za-z]+$', \"\", allele_1_n)\n",
    "                            genotype[gene][idx] = allele_1\n",
    "                            if genotype[gene][1] == '.':\n",
    "                                genotype[gene][1] = genotype[gene][0]\n",
    "                    \n",
    "                    elif self.method == 'Optitype':\n",
    "                        if 'Reads' not in cols:\n",
    "                            genes = [\"HLA-A\", \"HLA-B\", \"HLA-C\"]\n",
    "                            for g in genes:\n",
    "                                if (g not in genotype):\n",
    "                                    genotype[g] = [\".\", \".\"]\n",
    "                                if (g not in allGenes):\n",
    "                                    allGenes[g] = 1\n",
    "\n",
    "                            strG = 'HLA-'\n",
    "                            idx = 0\n",
    "                            for i in range(1,7):\n",
    "                                allele = strG + cols[i]\n",
    "                                gene = genes[idx]\n",
    "                                if i%2 == 1:\n",
    "                                    genotype[gene][0] = allele\n",
    "                                else:\n",
    "                                    genotype[gene][1] = allele\n",
    "                                    idx +=1\n",
    "            for gene in sorted(allGenes.keys()):\n",
    "                matchCnt = [0, 0, 0, 0]\n",
    "                if gene not in tgenotype:\n",
    "                    continue\n",
    "                elif tgenotype[gene] == []:\n",
    "                    continue\n",
    "                    #print(gene, \" not found in sample \", sname)\n",
    "                else:\n",
    "                    maxAllele = self.getMaxAllele(genotype[gene], tgenotype[gene])\n",
    "                    \n",
    "                    for k in range(2):\n",
    "                        if (genotype[gene][k] == maxAllele[k]):\n",
    "                            matchCnt[0] += 1\n",
    "                        else:\n",
    "                            matchCnt[2] += 1\n",
    "\n",
    "                    accuracy = 1\n",
    "                    weakAccuracy = 1\n",
    "                    total = sum(matchCnt[0:3])\n",
    "                    if (total > 0):\n",
    "                        accuracy = matchCnt[0] / total\n",
    "                        weakAccuracy = (matchCnt[0] + matchCnt[1]) / total\n",
    "                    s = \"\\t\".join([gene] + [\"%.4f\"%accuracy, \"%.4f\"%weakAccuracy] + [str(x) for x in matchCnt]) + \"\\n\"\n",
    "                    self.total_sum[gene].append(s)\n",
    "                    self.samplesOut[gene].append(sname)\n",
    "    \n",
    "    def writeOut(self):\n",
    "        \n",
    "        print(\"Start writing output file...\")\n",
    "        file = self.outpath + \"allele_level_summary.tsv\"\n",
    "        os.makedirs(os.path.dirname(self.outpath), exist_ok=True)\n",
    "        ppath = self.outpath + \"result_sum/\"\n",
    "        os.makedirs(os.path.dirname(ppath), exist_ok=True)\n",
    "        \n",
    "        print(\"Final output:\")\n",
    "        with open(file, \"w\") as outf:\n",
    "            for gene in self.total_sum.keys():\n",
    "                sample = self.samplesOut[gene]\n",
    "                match = 0\n",
    "                partmatch = 0\n",
    "                mismatch = 0\n",
    "                ignore = 0\n",
    "                strongMatch = 1\n",
    "                weakMatch = 1\n",
    "                index = 0\n",
    "                ffile = ppath + gene + \"_consistency.tsv\"\n",
    "                with open(ffile, \"w\") as out:\n",
    "                    for line in self.total_sum[gene]:\n",
    "                        s = sample[index]\n",
    "                        out.write(\"\\t\".join([s,line]))\n",
    "                        index +=1\n",
    "                        line = line.strip().split(\"\\t\")\n",
    "                        match += int(line[3])\n",
    "                        partmatch += int(line[4])\n",
    "                        mismatch += int(line[5])\n",
    "                        ignore += int(line[6])\n",
    "                    total = match + partmatch + mismatch\n",
    "                    if (total > 0):\n",
    "                        strongMatch = match / total\n",
    "                        weakMatch = (match + partmatch) / total\n",
    "                    finalOut = \"\\t\".join([gene, \"%.4f\"%strongMatch, \"%.4f\"%weakMatch , str(match), str(partmatch), str(mismatch), str(ignore)]) + \"\\n\"\n",
    "                    outf.write(finalOut)\n",
    "                    print(finalOut)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58f09515",
   "metadata": {},
   "source": [
    "### t1k hla rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "41f8d030",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  463  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9957\t0.9957\t922\t0\t4\t0\n",
      "\n",
      "HLA-B\t0.9946\t0.9946\t921\t0\t5\t0\n",
      "\n",
      "HLA-C\t0.9892\t0.9892\t916\t0\t10\t0\n",
      "\n",
      "HLA-DQB1\t0.9888\t0.9888\t706\t0\t8\t0\n",
      "\n",
      "HLA-DRB1\t0.9903\t0.9903\t917\t0\t9\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath)\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca68e251",
   "metadata": {},
   "source": [
    "### t1k hla wes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "242442e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  461  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9425\t0.9425\t869\t0\t53\t0\n",
      "\n",
      "HLA-B\t0.9805\t0.9805\t904\t0\t18\t0\n",
      "\n",
      "HLA-C\t0.9610\t0.9610\t886\t0\t36\t0\n",
      "\n",
      "HLA-DQB1\t0.9577\t0.9577\t680\t0\t30\t0\n",
      "\n",
      "HLA-DRB1\t0.9848\t0.9848\t908\t0\t14\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath)\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb81c279",
   "metadata": {},
   "source": [
    "### t1k hla wes whitelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "e6f582a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  461  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9816\t0.9816\t905\t0\t17\t0\n",
      "\n",
      "HLA-B\t0.9870\t0.9870\t910\t0\t12\t0\n",
      "\n",
      "HLA-C\t0.9805\t0.9805\t904\t0\t18\t0\n",
      "\n",
      "HLA-DQB1\t0.0000\t0.0000\t0\t0\t710\t0\n",
      "\n",
      "HLA-DRB1\t0.0000\t0.0000\t0\t0\t922\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath)\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5929542",
   "metadata": {},
   "source": [
    "### arcasHLA rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "a947622a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  463  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9309\t0.9309\t862\t0\t64\t0\n",
      "\n",
      "HLA-B\t0.9924\t0.9924\t919\t0\t7\t0\n",
      "\n",
      "HLA-C\t0.9773\t0.9773\t905\t0\t21\t0\n",
      "\n",
      "HLA-DQB1\t0.9468\t0.9468\t676\t0\t38\t0\n",
      "\n",
      "HLA-DRB1\t0.9924\t0.9924\t919\t0\t7\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath, method='arcasHLA')\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "912788be",
   "metadata": {},
   "source": [
    "### optitype rna "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "772a5d65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  463  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9914\t0.9914\t918\t0\t8\t0\n",
      "\n",
      "HLA-B\t0.9892\t0.9892\t916\t0\t10\t0\n",
      "\n",
      "HLA-C\t0.9687\t0.9687\t897\t0\t29\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath, gene = 'HLA', method = 'Optitype')\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81a71ee8",
   "metadata": {},
   "source": [
    "### optitype wes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "b1ce6fda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing current directories...\n",
      "Found  462  number of sample output!\n",
      "Reformatting and Storing ground truth...\n",
      "Start matching alleles...\n",
      "Start writing output file...\n",
      "Final output:\n",
      "HLA-A\t0.9913\t0.9913\t914\t0\t8\t0\n",
      "\n",
      "HLA-B\t0.9870\t0.9870\t910\t0\t12\t0\n",
      "\n",
      "HLA-C\t0.9620\t0.9620\t887\t0\t35\t0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "groundtruth = 'HLA_ground_truth.tsv'\n",
    "inPath = './'\n",
    "outPath = './'\n",
    "t1kResult = EvaluateT1k(inpath = inPath, gtrue=groundtruth, outpath=outPath, gene = 'HLA', method = 'Optitype')\n",
    "t1kResult.readDirs()\n",
    "t1kResult.setGtruth()\n",
    "t1kResult.matchAlleles()\n",
    "t1kResult.writeOut()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
